bson/src/main/org/bson/codecs/configuration/mapper/conventions/Converter.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright (c) 2008-2015 MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.bson.codecs.configuration.mapper.conventions;
+
+/**
+ * This defines a conversion between two types for use when
+ *
+ * @param <F> The "from" type as exists in the entity field being converted.
+ * @param <T> The "to" value which this Converter emits
+ */
+public interface Converter<F, T> {
+    /**
+     * Converts a value of type F to one of type T
+     *
+     * @param value the value to convert
+     * @return the new T value
+     */
+    T apply(F value);
+
+    /**
+     * @return The Class representing the type T
+     */
+    Class<T> getType();
+
+    /**
+     * Converts a value of type T back to one of type F
+     *
+     * @param value the value to convert
+     * @return the new F value
+     */
+    F unapply(T value);
```

Comment:
I'm open to changing the names if it'd help.  `apply()` felt the natural
choice which led to `unapply()` which, i'll admit, was a nod to Scala.  We
could change to `convert()`/`unconvert()`  or `deconvert()`.  I'm just not
sure what a decent name would be.

*--------------------------------*

*{ *name     : *"**Justin Lee**", *
  title    : *"*Software Engineer*"*,
  twitter  : *"*@evanchooly <http://twitter.com/evanchooly>*"*,
  web      : *[* "10gen.com", *"*antwerkz.com*"* *]*,
  location : *"*New York, NY*"* *}*

On Tue, Jul 14, 2015 at 9:12 AM, Ross Lawley <notifications@github.com>
wrote:

> In
> bson/src/main/org/bson/codecs/configuration/mapper/conventions/Converter.java
> <https://github.com/mongodb/mongo-java-driver/pull/323#discussion_r34565723>
> :
>
> > +     * @return the new T value
> > +     */
> > +    T apply(F value);
> > +
> > +    /**
> > +     * @return The Class representing the type T
> > +     */
> > +    Class<T> getType();
> > +
> > +    /**
> > +     * Converts a value of type T back to one of type F
> > +     *
> > +     * @param value the value to convert
> > +     * @return the new F value
> > +     */
> > +    F unapply(T value);
>
> apply is fine as it can return type T, its more the unapply convention of
> returning Option[F] rather than a flat F.
>
> I'm not sure it has any real implications, other than perhaps surprising
> to some Scala users. It could mean you have to use a class rather than an
> object when creating custom Converters.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mongodb/mongo-java-driver/pull/323/files#r34565723>.
>